iceoryx2-bb/posix/src/directory.rs

```diff
@@ -379,64 +380,76 @@ impl Directory {
         let number_of_directory_entries =
             unsafe { posix::scandir(self.path.as_c_str(), &mut namelist) };
 
-        if number_of_directory_entries < 0 {
-            let msg = "Unable to read directory contents";
-            handle_errno!(DirectoryReadError, from self,
-                Errno::EACCES => (InsufficientPermissions, "{} due to insufficient permissions.", msg),
-                Errno::ENOENT => (DirectoryDoesNoLongerExist, "{} since the directory does not exist anymore.", msg),
-                Errno::ENOMEM => (InsufficientMemory, "{} due to insufficient memory.", msg),
-                Errno::EMFILE => (PerProcessFileHandleLimitReached, "{} since the file descriptor limit of the process was reached.", msg),
-                Errno::ENFILE => (SystemWideFileHandleLimitReached, "{} since the system-wide limit of file descriptors was reached.", msg),
-                v => (UnknownError(v as i32), "{} since an unknown error occurred ({}).", msg, v)
-            );
-        } else {
-            let mut contents: Vec<DirectoryEntry> = vec![];
-            for i in 0..number_of_directory_entries {
-                let raw_name = unsafe {
-                    (*(*namelist.offset(i as isize))).d_name.as_ptr() as *mut posix::c_char
-                };
-                let raw_name_length = unsafe { strnlen(raw_name, FileName::max_len()) };
-
-                const DOT: posix::c_char = b'.' as _;
-                // dot is skipped
-                if raw_name_length == 1 && unsafe { *raw_name == DOT } {
-                    continue;
+        let _memory_cleanup_guard = ScopeGuardBuilder::new(namelist)
```

Comment:
Nice, something like this would also be great for the C bindings but I need to discard the drop on the happy path. Need to look at the API how easily this can be done. Maybe we need something which builds upon the `ScopeGuard`, e.g. `CleanupGuard` which has an `ok()` (or similar) method to discard the cleanup callback